Sources/UrlFormEncoding/UrlFormDecoder.swift

```diff
@@ -0,0 +1,629 @@
+  }
+}
+
+extension UrlFormDecoder.UnkeyedContainer.Key: CodingKey {
+  public var stringValue: String {
+    return String(self.index)
+  }
+
+  public init?(stringValue: String) {
+    guard let intValue = Int(stringValue) else { return nil }
+    self.init(intValue: intValue)
+  }
+
+  public var intValue: Int? {
+    return .some(self.index)
+  }
+
+  public init?(intValue: Int) {
+    self.init(index: intValue)
+  }
+}
+
+private let iso8601: (DateFormatter) -> DateFormatter = {
+  $0.calendar = Calendar(identifier: .iso8601)
+  $0.locale = Locale(identifier: "en_US_POSIX")
+  $0.timeZone = TimeZone(abbreviation: "UTC")
+  return $0
+}
+
+private let iso8601DateFormatter = DateFormatter()
+  |> iso8601
+  |> \.dateFormat .~ "yyyy-MM-dd'T'HH:mm:ss.SSSXXXXX"
+
+private let iso8601DateFormatterWithoutMilliseconds = DateFormatter()
+  |> iso8601
+  |> \.dateFormat .~ "yyyy-MM-dd'T'HH:mm:ssXXXXX"
+
+private func parse(isArray: @escaping (String) -> Bool) -> (String) -> [String: Any] {
+  func parseHelp(_ params: inout [String: Any], _ path: [String], _ value: Any) {
+    let key = path[0]
+
+    if path.count == 1 {
+      params[key] = value
+    } else if path.count == 2 && isArray(path[1]) {
+      var values = params[key] as? [Any] ?? []
+      values.append(value)
+      params[key] = values
+    } else if isArray(path[1]) {
+      var (values, nested) = (params[key] as? [Any] ?? [], [:] as [String: Any])
+      parseHelp(&nested, Array(path[2...]), value)
+      values.append(nested)
+      params[key] = values
+    } else {
+      var values = params[key] as? [String: Any] ?? [:]
+      parseHelp(&values, Array(path[1...]), value)
+      params[key] = values
+    }
+  }
+
+  return { query in
+    var params: [String: Any] = [:]
+
+    for (name, value) in pairs(query) {
+      let result = name.reduce(into: (path: [] as [String], current: "")) { result, char in
+        switch char {
+        case "[":
+          if result.path.isEmpty {
+            result.path.append(result.current)
+            result.current.removeAll()
+          }
+        case "]":
+          result.path.append(result.current)
+          result.current.removeAll()
+        default:
+          result.current.append(char)
+        }
+      }
+      let path = result.current.isEmpty ? result.path : result.path + [result.current]
+      parseHelp(&params, path.isEmpty ? [""] : path, value)
+    }
+
+    return params
+  }
+}
+
+private func pairs(_ query: String) -> [(String, String)] {
+  return query
+    .split(separator: "&")
+    .map { (pairString: Substring) -> (name: String, value: String) in
+      let pairArray = pairString.split(separator: "=", maxSplits: 1, omittingEmptySubsequences: false)
+        .flatMap { String($0).removingPercentEncoding }
+      return (pairArray[0], pairArray.count == 2 ? pairArray[1] : "")
+    }
+    .sorted { $0.name < $1.name }
+}
+
+private func singleton(_ value: Any) -> String? {
```

Comment:
Oh, one other thing in this file! It seems safest (and was easier to implement) to decode `A` or `[A]` depending on the data structure without worrying about the query strategy. Happy to discuss any concerns, though!

Reply:
not sure I understand. could you explain a bit more?

Reply:
Ah, an example would help! Given the following struct:

``` swift
struct Box: Decodable {
  let a: [Int]
  let b: Int
}
```

And the following query string:

```
a=1&b=1
```

The underlying data representation is `["a": ["1"], "b", ["1"]]`, so `b` can unbox the value using this helper. It's mostly a concession for the default representation, though, so maybe its logic should move to the strategy somehow?